### DISCUSSION TITLE
input and params  specified as string names

### DISCUSSION THREAD

**Discussion by apostopher** [statelyai/xstate#4757] [Q&A]
the `input` for invoked actors and `params` for actions and guards can only be specified inline.
it will be good to support `named input` and `named params`

e.g.

## Specify named input in the invoke config.
```
// example state
Loading: {
      invoke: {
        id: "fetchData",
        input: "fetchDataInput", // this is provided at runtime using `setup` function.
        onDone: {
          target: "Success",
          actions: {
            type: "setData",
            params: "setDataParams"
          },
        },
        onError: {
          target: "Failure",
          actions: {
            type: "setError",
          },
        },
        src: "inline:fetchData.Loading#actor[0]",
      },
      description:
        "The machine is currently loading the data based on the provided input.",
    }
```

## Provide implementation of the named `input` & `params` in setup function / machine config
```
setup({
  types: {…},
  actions: {…},
  actors: {...},
  inputs: {
    fetchDataInput: ({ context }) => ({ accessToken: context.accessToken })
  },
  params: {
    setDataParams: { format: 'csv' }
  }
})
```

## Why?
by allowing named inputs and params, the machine config is serialisable. we can send it via api, store it in db, download as json etc.

**Comment by Andarist** [comment]
We'll consider that but it could put a lot of pressure on TypeScript here since we can only allow compatible strings. This would require O(m * n) relationship checks.

**Reply by apostopher** [reply]
Thanks! got it. i'll look into the current implementation for `named actions and guards` to get a better understanding.
